server/repository/src/db_diesel/invoice_line_row.rs

```diff
@@ -97,14 +101,36 @@ impl<'a> InvoiceLineRowRepository<'a> {
         InvoiceLineRowRepository { connection }
     }
 
-    pub fn upsert_one(&self, row: &InvoiceLineRow) -> Result<(), RepositoryError> {
+    pub fn upsert_one(&self, row: &InvoiceLineRow) -> Result<i64, RepositoryError> {
         diesel::insert_into(invoice_line)
             .values(row)
             .on_conflict(id)
             .do_update()
             .set(row)
             .execute(self.connection.lock().connection())?;
-        Ok(())
+        self.insert_changelog(row, RowActionType::Upsert)
+    }
+
+    fn insert_changelog(
+        &self,
+        row: &InvoiceLineRow,
+        action: RowActionType,
+    ) -> Result<i64, RepositoryError> {
+        let invoice = InvoiceRowRepository::new(self.connection).find_one_by_id(&row.invoice_id)?;
+        let invoice = match invoice {
```

Comment:
My opinionated stylistic preference, not asking for change, happy with match

`invoice.ok_or(RepositoryError::NotFound)?`

Reply:
That's is nice! Thanks for the reminder. I tend to revert to the pattern I know, but very happy with that approach too.